src/hotspot/share/opto/divconstants.cpp

```diff
@@ -0,0 +1,209 @@
+//
+// We have x / d <= x * c / m < (x * ((v + 1) / v)) / d
+//
+// As a result, since x * ((v + 1) / v) <= x * ((x + 1) / x) = x + 1
+//
+// x / d <= x * c / m < (x + 1) / d, which implies floor(x / d) = floor(x * c / m) since
+// there can be no integer in (x / d, (x + 1) / d)
+//
+// For v + 1 <= x <= v + d - 1, since v >= d - 1, we have x <= 2v
+// As a result, x * ((v + 1) / v) <= x * ((x + 2) / x)
+//
+// floor(x / d) = (v + 1) / d
+// x * c / m < x * ((v + 1) / v) / d) <= x * ((x + 2) / x) / d = (x + 2) / d
+//           <= (v + d - 1 + 2) / d = (v + 1) / d + 1
+// Which means (v + 1) / d <= floor(x * c / m) < (v + 1) / d + 1 with (v + 1) / d being an integer
+// This implies floor(x / d) = floor(x * c / m) for v + 1 <= x <= v + d - 1
+//
+// Combining all the cases gives us the conclusion.
+//
+// (b) Since ceil(a / b) = floor((a - 1) / b) + 1, we need to prove:
+//
+// floor((x - 1) / d) = floor(x * c / m)
+//
+// For 0 > x >= -v
+//
+// Since 1 / d < c / m <= (1 / d) * ((v + 1) / v)
+//
+// We have x / d > x * c / m >= (x / d) * ((v + 1) / v)
+//
+// since x * ((v + 1) / v) >= x * ((x - 1) / x) = x - 1
+// x / d > x * c / m >= (x - 1) / d, which implies floor((x - 1) / d) = floor(x * c / m) since
+// there can be no integer in ((x - 1) / d, x / d)
+//
+// For -v - d + 1 <= x <= -v - 1, since v >= d - 1, we have x >= -2v
+// As a result, x * ((v + 1) / v) >= x * ((x - 2) / x) = x - 2
+//
+// x / d <= (-v - 1) / d
+// floor((x - 1) / d) = (-v - 1) / d - 1
+// x * c / m >= x * ((v + 1) / v) / d) >= (x - 2) / d
+//           >= (-v - d + 1 - 2) / d = (-v - 1) / d - 1
+// which means (-v - 1) / d >= x / d > x * c / m >= (-v - 1) / d - 1
+// This implies floor((x - 1) / d) = floor(x * c / m) for -v - 1 >= x >= -v - d + 1
+//
+// Combining all the cases gives us the conclusion.
+//
+// (3) Discussion:
+//
+// Let x be v, v - d + 1, -v, -v + d - 1, it can be seen that these bounds are indeed optimal
+//
+// (4) Implementation:
+//
+// For computation efficiency, we only consider the values m = 2**s.
+// This function does not handle the cases d being a power of 2, which means
+// that c * d is never equal to m.
+//
+// We find the value of c, m such that it satisfies the bounds for both the nonnegative
+// and negative ranges of x. This can be done by finding v_neg and v_pos and the bounds of
+// c * d - m is the intersection of (0, m / v_neg] and (0, m / v_pos). Which is (0, m / v_pos)
+// if v_pos >= v_neg and (0, m / v_neg] otherwise.
+//
+// Given v = max(v_neg, v_pos). The function inductively calculates c, rc, qv, rv such that:
+//
+// c * d - rc = 2**s with 0 < rc <= d
+// qv * v + rv = 2**s with 0 <= rv < v
+template <class T>
+void magic_divide_constants(T d, T N_neg, T N_pos, juint min_s, T& c, bool& c_ovf, juint& s) {
+  static_assert(std::is_unsigned<T>::value, "calculations must be done in the unsigned domain");
+  assert(!is_power_of_2(d), "this case should be handled separately");
```

Comment:
The algorithm also works for `d` a power of 2 (except, perhaps, `d = 1`).
But of course, it makes more sense to handle them separately.